diff -Naru curl-7.16.1/lib/connect.c curl-7.16.1-mod/lib/connect.c
--- curl-7.16.1/lib/connect.c	2006-12-22 15:30:54.000000000 +0200
+++ curl-7.16.1-mod/lib/connect.c	2007-03-06 13:58:34.000000000 +0200
@@ -261,7 +261,9 @@
     char myhost[256] = "";
     in_addr_t in;
     int rc;
-    bool was_iface = FALSE;
+    char ipv6_addr[16];
+    bool was_iface = FALSE; 
+    int in6 = -1;
 
     /* First check if the given name is an IP address */
     in=inet_addr(data->set.device);
@@ -342,7 +344,11 @@
 #endif
 
     in=inet_addr(myhost);
-    if (CURL_INADDR_NONE == in) {
+
+#ifdef ENABLE_IPV6
+    in6 = inet_pton (AF_INET6, myhost, (void *)&ipv6_addr);
+#endif
+    if (CURL_INADDR_NONE == in && -1 == in6) {
       failf(data,"couldn't find my own IP address (%s)", myhost);
       return CURLE_HTTP_PORT_FAILED;
     } /* end of inet_addr */
