--- lib/multi.c 23 Mar 2007 22:25:04 -0000      1.133
+++ lib/multi.c 24 Mar 2007 00:29:53 -0000
@@ -1488,6 +1488,6 @@
       return NULL; /* no messages left to return */
     easy=multi->easy.next;
-    while(easy) {
+    while(easy != &multi->easy) {
       if(easy->msg_num) {
         easy->msg_num--;
         break;
@@ -1999,7 +1999,7 @@
   int i;
   fprintf(stderr, "* Multi status: %d handles, %d alive\n",
           multi->num_easy, multi->num_alive);
-  for(easy=multi->easy.next; easy; easy = easy->next) {
+  for(easy=multi->easy.next; easy != &multi->easy; easy = easy->next) {
     if(easy->state != CURLM_STATE_COMPLETED) {
       /* only display handles that are not completed */
       fprintf(stderr, "handle %p, state %s, %d sockets\n",
